File: Python/69.sqrt-x.py
#
# @lc app=leetcode id=69 lang=python3
#
# [69] Sqrt(x)
#

# @lc code=start
class Solution:
    def mySqrt(self, x: int) -> int:
        l, r = 0, x
        
        while l <= r:
            mid = (l + r) // 2
            
            if mid * mid > x:
                r = mid -1
            elif mid * mid < x:
                l = mid +1
            else:
                return mid
        
        return r
    
# @lc code=end

File: TypeScript/69.sqrt-x.ts
/*
 * @lc app=leetcode id=69 lang=typescript
 *
 * [69] Sqrt(x)
 */

// @lc code=start
function mySqrt(x: number): number {
    if (x === 1) return x

    let l: number = 0
    let r: number = x
    let mid: number = 0

    while (l <= r) {
        mid = Math.floor((l + r) / 2)

        if (mid * mid > x) {
            r = mid - 1
        } 
        else if (mid * mid < x){
            l = mid + 1
        } else {
            return mid
        }
    }

    return r
};
// @lc code=end

